File: 1873 E. Building an Aquarium/Solution.c
// 315916811	        Apr/17/2025 12:25UTC+6	        Krishno_MD	        1873E - Building an Aquarium	        GNU C11	        Accepted	        155 ms	        900 KB

#include<stdio.h>
#include<stdlib.h>

// Krishno_MD_KUET_CSE

void solution(){
    long long n,x; scanf("%lld %lld",&n,&x);
    int v[n];
    for(long long i=0; i<n; i++){scanf("%lld",&v[i]);}
    long long ans=-1,hs=1,he=1e10;
    while(hs<=he){
        long long w=0, mid = hs+(he-hs)/2;
        for(long long i=0; i<n; i++){
            if(mid>v[i]){w += (mid-v[i]);}
        }
        if(w<=x){ans=mid; hs=mid+1;}
        else{he=mid-1;}
    }
    printf("%lld\n",ans);
}

int main(){
    int t;
    scanf("%d",&t);
    while(t--){
        solution();
    }
    return 0;
}

File: 1873 E. Building an Aquarium/Solution.py
# 315916992	        Apr/17/2025 12:27UTC+6	        Krishno_MD	        1873E - Building an Aquarium	        PyPy 3-64	        Accepted	        265 ms	        23000 KB

# Krishno_MD_KUET_CSE

t = int(input())

for i in range(t):
    n,x = map(int,input().split())
    v = list(map(int,input().split()))
    ans,hs,he = -1,1,1e10
    while hs<=he:
        w,mid = 0, hs+int((he-hs)/2)
        for i in v:
            if mid>i:
                w += (mid-i)
        if w<=x:
            ans=mid
            hs=mid+1
        else:
            he=mid-1
    print(ans)
